share/doc/wake/logging.md

```diff
@@ -0,0 +1,64 @@
+# Logging
+
+This article gives a high level overview of how wake logging works.
+Check out [print.wake](https://github.com/sifive/wake/blob/master/share/wake/lib/core/print.wake)
+for commented code.
+
+## How Logging works on Wake Execution commands
+
+Every `Job` run by wake will have its own stdout and stderr, and when the underlying process prints to stdout or stderr, it will go the `Job`-specific stdout and stderr, not the stdout and stderr of the `wake` process itself.
+
+When writing your wake code, you can control whether these `Job`-specific stdout and stderr are also printed to `wake`'s own stdout, stderr, or other arbitrary file descriptors. These can be controlled with `setPlanStderr` and `setPlanStdout` before you `runJob`.
+We call this which `stream` each goes to (we often refer to a `stream` as a `logLevel`).
+
+Regardless of which `stream` you direct the `Job`'s `Stderr` and `Stdout`,
+`wake` will store all the text in its database,
+discarding the information about the `stream` it was destined for at the time it was stored.
+It will however know whether it came from the `Job`'s stderr/stdout.
+
+So what's the difference between the different streams?
+What you see on your console when you run wake.
+It controls *what* you see (depending on the command line flags) and what it looks like (what color it is).
+It also controls which *actual* stream it is going to coming out of `wake`
+(stderr or stdout, or other file descriptors 3-5).
+
+There are a number of built in streams in wake which all start with `log...`.
+You can also make your own stream using `mkLogLevel`.
+You can use this to customize the color you want it to be shown as,
+or (more commonly) to grab output from a `Job` for passing to another `Job`,
+consuming within your wake program, or redirecting outside of wake using shell redirect.
```

Comment:
what about the "consuming within your wake program" use case?

Reply:
I don't actually understand what you mean by that.

Reply:
like, can i do something in wake like send some Job's stdout to "megalog" (`def logMega = mkLogLevel "megalog"`, then send a job's stdout to `logMega`), then do some sort of in-wake decision making on the text found in `logMega`.

Reply:
In other words, is the only way to actually use the custom case shell redirect? If so I should just say that and then maybe give a little cookbook at the end. If there are other use cases like consuming the output within wake, then maybe that's another cookbook or I'd leave out all the cookbooks

Reply:
> like, can (should?) I do something in wake like send some Job's stdout to "megalog" (def logMega = mkLogLevel "megalog", then send a job's stdout to logMega), then do some sort of in-wake decision making on the text found in logMega.

That's not possible within wake itself.

> In other words, is the only way to actually use the custom case shell redirect? If so I should just say that and then maybe give a little cookbook at the end. If there are other use cases like consuming the output within wake, then maybe that's another cookbook or I'd leave out all the cookbooks

No and yes. Yes if you want to make sure nothing else is written to stdout and stderr, which is important when you want to consume the content within the custom stream programmatically. No if you don't care, in which case you can just redirect one of the builtin streams (loggers?) if you wanted to.

The shell redirection thing is a shell-specific thing. You can consume the custom streams without redirection if you directly read the file descriptors, such as if you invoked wake via another programming language like Python.